Use docker-compose to run Solr in CI

mjgiarlo opened the pull request:
Fixes #2195
Fixes #2231 
Closes #2256 

# Summary

In this way, we are using docker-compose for development and testing of the Blacklight gem itself, not for downstream purposes such as generation of new Blacklight applications, which will continue to use solr_wrapper. Also includes: using a lighter base image for the app.

This supersedes #2256.

# Open Question (Now Closed)

I put the documentation here: https://github.com/projectblacklight/blacklight/wiki/Testing-and-Developing-Blacklight

~~It is not clear to me where information like this is currently documented. The README is short and sweet, and for this reason I suspect it is not a good place for such internal development documentation to live. Other options include `CONTRIBUTING.md`, but that is more focused on higher-level contribution process than particular technical steps.~~

~~Note, I also don't think the Quickstart is the right place, so this ticket has nothing to do with #2183, which I think we should either close as a `wontfix` per [my comment](https://github.com/projectblacklight/blacklight/issues/2183#issuecomment-598779218) or discuss what the right way to move forward is particularly w/r/t how much we continue to generate for downstream BL apps: *i.e.*, if we don't give them a working SolrWrapper, do we generate/ship a working docker-compose configuration? Or do we get out of that business entirely? This is beyond the scope of this PR.~~

~~For this reason, perhaps the best place is somewhere on the [project wiki](https://github.com/projectblacklight/blacklight/wiki). Happy to take suggestions on this.~~

# Webpack(er) Broken (But Not Related)

I believe this may be solved by https://github.com/projectblacklight/blacklight/pull/2266. Thanks, @mejackreed!

~~Note that this PR changes the `blacklight:server` rake task to use `docker-compose` for Solr but I was unable to see this working---with Ruby 2.6 and Rails 6, but other combinations will yield different results---because when Blacklight uses EngineCart to generate the test app, there's a problem with webpack(er). If you generate without the `--skip-webpack-install` flag being passed to the `rails new` command, the Rails webpack generator runs and fails because the dependency is not declared. If you skip webpack installation, Rails generates an app that still contains `webpacker` in its Gemfile, and it's a multi-step process to strip out webpack(er) and use Sprockets in its stead ([1](http://mitrev.net/ruby/2019/12/27/rails-6-without-webpack/), [2](https://www.reddit.com/r/rails/comments/b0goct/webpacker_for_non_js_people/eih2930/), [3](https://stackoverflow.com/a/58518795)). That said, I did not alter this behavior or attempt to fix the underlying problem since it is orthogonal to how we spin up Solr for CI/testing of Blacklight.~~

mjgiarlo commented on .travis.yml:
```diff
@@ -1,10 +1,12 @@
   chrome: stable
-language: ruby
 
-notifications:
```
This is dead configuration, overridden by lines 34-37 below, and has been cruft for 7 years.

mjgiarlo commented on .travis.yml:
```diff
@@ -1,10 +1,12 @@
 
-notifications:
-  email: false
+services:
```
This may not be needed here. If the current builds pass, I'll try yanking it out and re-pushing.

mjgiarlo commented on .travis.yml:
```diff
@@ -35,16 +36,16 @@ notifications:
-
-jdk: openjdk11
+env:
+  global:
```
This appears to be the more modern way to specify global environment variables.

mjgiarlo commented on .travis.yml:
```diff
@@ -35,16 +36,16 @@ notifications:
-  - ENGINE_CART_RAILS_OPTIONS='--skip-git --skip-listen --skip-spring --skip-keeps --skip-action-cable --skip-coffee --skip-test'
-  - CC_TEST_REPORTER_ID=5042c7358c96b0b926088a4cda3e132fffe7a66ce8047cdb1dc6f0b4b6676b79
-
-jdk: openjdk11
```
Don't need JDK anymore if spinning up Solr via Travis's docker infrastructure and our docker-compose configuration.

mjgiarlo commented on tasks/blacklight.rake:
```diff
@@ -25,42 +30,31 @@ namespace :blacklight do
     Rake::Task["spec"].invoke
   end
 
-  desc "Create the test rails app"
```
Unclear why this is needed; looks like cruft from the pre-EngineCart days. Found no references to it anywhere on the open web, so I removed it.

dkinzer replied:
I think this is needed because blacklight is a rails engine too and to test that functionality the engine_cart is used.

mjgiarlo replied:
Removing this doesn't seem to have caused any ill effects, FWIW.

mjgiarlo commented on tasks/blacklight.rake:
```diff
@@ -25,42 +30,31 @@ namespace :blacklight do
     task seed: ['engine_cart:generate'] do
       within_test_app do
-        system "bundle exec rake blacklight:index:seed"
+        system "bin/rake blacklight:index:seed"
```
Use generated binstubs.

dkinzer commented on docker-compose.yml:
```diff
@@ -14,17 +14,19 @@ services:
       - $PWD/lib/generators/blacklight/templates/solr/conf:/opt/solr/conf
     ports:
-      - "8983:8983"
+      - "${SOLR_PORT}:8983"
```
I really like the use of SOLR_PORT environment variable here.

mjgiarlo replied:
Thanks!

dkinzer commented:
@mjgiarlo I wonder whether (if it's possible) for us to also incorporate the docker app in the travis build in order to avoid issues where the Dockerfile gets broken or stale.

mjgiarlo commented:
Totally open to that, @dkinzer. Do you have any ideas for what to test? Could try to build the image and fail unless it returns a zero exit code? Or maybe there's a smarter thing to do?

dkinzer commented:
Maybe we can run the blacklight tests within the app for one of the ruby versions?

Unfortunately when I try that on my local I get an error so it's likely not an easy thing to setup:

```
 docker-compose exec app rake ci
WARN: Unresolved or ambigious specs during Gem::Specification.reset:
      minitest (~> 5.1)
      Available/installed versions of this gem:
      - 5.14.0
      - 5.13.0
      - 5.11.3
      loofah (~> 2.3)
      Available/installed versions of this gem:
      - 2.4.0
      - 2.3.1
      nokogiri (>= 1.6)
      Available/installed versions of this gem:
      - 1.10.9
      - 1.10.5
      builder (~> 3.1)
      Available/installed versions of this gem:
      - 3.2.4
      - 3.2.3
WARN: Clearing out unresolved specs. Try 'gem cleanup <gem>'
Please report a bug if this causes problems.
You must `gem install bundler` and `bundle install` to run rake tasks
rake aborted!
LoadError: cannot load such file -- engine_cart/rake_task
tasks/blacklight.rake:3:in `require'
tasks/blacklight.rake:3:in `<top (required)>'
/app/rakefile:13:in `load'
/app/rakefile:13:in `<top (required)>'
/usr/local/bundle/gems/rake-13.0.1/exe/rake:27:i```

dkinzer commented:
I was able to run the tests using `docker-compose exec app bundle exec rspec` but now that I think about it, I tend to agree with you that just doing `docker-compose up` should suffice.

mjgiarlo commented on bin/start.sh:
```diff
@@ -1,5 +0,0 @@
-#!/usr/bin/env bash
```
I essentially moved this file into `.docker/app/` and renamed it to make its intent clearer.

cdmo commented on .docker/app/Dockerfile:
```diff
@@ -1,12 +1,24 @@
-      nodejs yarn build-essential libpq-dev \
-      && apt-get clean
-RUN gem install bundler
+FROM ruby:2.6.5-alpine
```
`docker` noob ✋  

I've got a couple questions because I'm just trying to learn.

So why 2.6.5-alpine? Specifically the `alpine` part. There are [a lot of flavors to choose from on docker hub](https://hub.docker.com/_/ruby). Also, why 2.6.5 and not some other version? 2.7 is the current stable release and Blacklight's docs say it can be run on ruby 2.2+.

Also wondering if this could be set to respect `.ruby-version` or is that missing the point?

dkinzer replied:
alpine is the better choice ... we're trying to switch to that at TU too.  It's because alpine images are smaller and they are more secure.. .so using them in production is smarter.

cdmo replied:
Guess using them everywhere is smarter too then, right? Or should I not make that assumption. Thanks.

mjgiarlo replied:
I don't believe there is a way to read the `.ruby-version` file, and I don't think Blacklight ships with a `.ruby-version` file.

I'll push a change momentarily that allows overriding the version of the Alpine Ruby base image via environment variable, defaulting to 2.6.5 (because 2.6 is what was in here previously, and 2.6.5 is the specific 2.6 version we are currently testing, but it's arbitrary). Sufficient?

mjgiarlo replied:
@cdmo :speech_balloon: 
> Guess using them everywhere is smarter too then, right? Or should I not make that assumption. Thanks.

Yeah, we use Alpine-based images for dev/test generally as well because they're lighter.

cdmo commented:
@mjgiarlo and @dkinzer do you have any recommendations for learning more about docker, docker compose, etc? Book? Good docs somewhere? Video? Thanks.

Anything that you need tested by developers here at the moment?

cdmo commented:
This will need to be rebased too I think to get the latest changes that were merged after this was created.

dkinzer commented:
@cdmo my resource of choice are the docker docs (dark mode theme turned on of course): https://docs.docker.com/compose/

jcoyne commented:
@cdmo some resources I recommend:
<https://LearnDocker.online>: Free resource with 11+ hours of video content teaching you everything about Docker and Containers.
<https://RailsWithDocker.com>: Free online resource teaching you how to dockerize a Rails app.

cdmo commented:
Slick. Thanks @jcoyne

mjgiarlo commented:
@cdmo :speech_balloon: 
> @mjgiarlo and @dkinzer do you have any recommendations for learning more about docker, docker compose, etc? Book? Good docs somewhere? Video? Thanks.
> 
> Anything that you need tested by developers here at the moment?

I'm very nearly ready to mark this PR as ready for review but the big need I see is some shared sense of where to document this stuff. See the `Open Question` section above. I welcome your thoughts!

mjgiarlo commented on .docker/app/Dockerfile:
```diff
@@ -1,12 +1,26 @@
-      nodejs yarn build-essential libpq-dev \
-      && apt-get clean
-RUN gem install bundler
+ARG ALPINE_RUBY_VERSION
```
I'm using `ALPINE_RUBY_VERSION` because `RUBY_VERSION` is already in use elsewhere, e.g., by RVM. This allows switching the version of Ruby used in the app image via environment variable:

```shell
$ ALPINE_RUBY_VERSION=2.7 docker-compose up --build app
```

mjgiarlo commented on .env:
```diff
@@ -0,0 +1,5 @@
+ALPINE_RUBY_VERSION=2.6.5
```
These are default values. All of them are overrideable via environment variable and will be used by the docker images.

cdmo replied:
Cool.

mjgiarlo commented on .travis.yml:
```diff
@@ -35,16 +33,17 @@ notifications:
+
 after_script:
   - ./cc-test-reporter after-build --exit-code $TRAVIS_TEST_RESULT
+  - docker-compose build app || travis_terminate 1 # validates application Dockerfile
```
We don't need to use the `app` image in the CI build, but we do want to be sure the image *does* build. This step validates that happens without errors. We want to do this after the build has run, since this will create `.internal_test_app` with root permissions which trips up the CI task.

mjgiarlo commented on tasks/blacklight.rake:
```diff
@@ -11,10 +11,15 @@ require 'rubocop/rake_task'
+
   within_test_app do
-    system "RAILS_ENV=test rake blacklight:index:seed"
+    system "RAILS_ENV=test bin/rake blacklight:index:seed"
```
Use generated binstubs. See https://github.com/rails/rails/blob/v6.0.2.1/railties/lib/rails/app_loader.rb#L13-L14

Not necessary for this work, so I'm happy to undo this if folks would prefer.

mjgiarlo commented on tasks/blacklight.rake:
```diff
@@ -11,10 +11,15 @@ require 'rubocop/rake_task'
 desc "Run test suite"
-task ci: ['blacklight:generate'] do
+task :ci do
+  system "docker-compose up -d solr"
```
Adding this means folks can pull down the Blacklight repo and launch `rake ci` and watch it work, ideally. @cdmo assuming you have docker-compose installed. Can you pull this branch and verify this? If not true, happy to iterate.

mejackreed replied:
I ran into an issue where I pulled this branch and all of the feature tests failed for me. `bundle exec rake ci`

mejackreed replied:
Also, I noticed that this is both in the `ci` task and the travis file. Is this needed in both placed?

cdmo replied:
I had the same experience as @mejackreed, feature specs failed:

`Finished in 24.6 seconds (files took 4.51 seconds to load)
1140 examples, 71 failures, 6 pending`

Looks like it's unrelated to this PR though if you ask me. I've seen this error before:

```
  64) Facets retains the filters when we change per page number
      Failure/Error: //= link_tree ../images

      ActionView::Template::Error:
        link_tree argument must be a directory
      # ./.internal_test_app/app/assets/config/manifest.js:1
      # ------------------
      # --- Caused by: ---
      # Sprockets::ArgumentError:
      #   link_tree argument must be a directory
      #   ./.internal_test_app/app/assets/config/manifest.js:1
```

I think this is a fix coming #2266 (pretty sure this is caused by sprockets 4.0.0 requiring a change https://github.com/rails/sprockets-rails/issues/443#issuecomment-540696186)

cdmo replied:
@mejackreed is travis just running `bundle exec rake`? I think I was expecting to see a `script` section and didn't, so I guess travis just has that [as a default](https://travis-ci.org/github/projectblacklight/blacklight/jobs/663558443#L470)?

cdmo replied:
I merged in the fix from #2266 and re-ran `ci` and got a new error:

```
  70) Bookmarks navigating from the homepage has a link to the history page
      Failure/Error: <%= javascript_include_tag "application" %>

      ActionView::Template::Error:
        Asset `application.js` was not declared to be precompiled in production.
        Declare links to your assets in `app/assets/config/manifest.js`.

          //= link application.js
        and restart your server
      # ./app/views/layouts/blacklight/base.html.erb:15:in `block in ___sers_cdm____esktop_blacklight_app_views_layouts_blacklight_base_html_erb__2593207400865544889_70245572263500'
      # ./app/views/layouts/blacklight/base.html.erb:2:in `___sers_cdm____esktop_blacklight_app_views_layouts_blacklight_base_html_erb__2593207400865544889_70245572263500'
      # ./app/views/layouts/blacklight.html.erb:17:in `___sers_cdm____esktop_blacklight_app_views_layouts_blacklight_html_erb__193081276090225314_70245569552620'
      # ./spec/support/features/session_helpers.rb:17:in `sign_in'
      # ./spec/features/bookmarks_spec.rb:6:in `block (3 levels) in <top (required)>'
      # ------------------
      # --- Caused by: ---
      # Sprockets::Rails::Helper::AssetNotPrecompiled:
      #   Asset `application.js` was not declared to be precompiled in production.
      #   Declare links to your assets in `app/assets/config/manifest.js`.
      #
      #     //= link application.js
      #   and restart your server
      #   ./app/views/layouts/blacklight/base.html.erb:15:in `block in ___sers_cdm____esktop_blacklight_app_views_layouts_blacklight_base_html_erb__2593207400865544889_70245572263500'
```

mjgiarlo replied:
@mejackreed 
> Also, I noticed that this is both in the `ci` task and the travis file. Is this needed in both places?

I'll try removing the one in `.travis.yml`

mjgiarlo replied:
> @mejackreed is travis just running `bundle exec rake`? I think I was expecting to see a `script` section and didn't, so I guess travis just has that [as a default](https://travis-ci.org/github/projectblacklight/blacklight/jobs/663558443#L470)?

@cdmo yep, that's what is happening, and the default rake task is this `ci` task

mjgiarlo replied:
@cdmo I can get the tests to pass A-OK in this branch. Here is the guide I just wrote up for how I did this:

https://github.com/projectblacklight/blacklight/wiki/Testing-and-Developing-Blacklight

mjgiarlo commented on tasks/blacklight.rake:
```diff
@@ -42,25 +44,15 @@ namespace :blacklight do
       Rake::Task['engine_cart:generate'].invoke
     end
 
-    SolrWrapper.wrap do |solr|
```
Removing the last vestiges of SolrWrapper used for internal testing.

mjgiarlo commented:
@dkinzer :speech_balloon: 
> @mjgiarlo I'm getting the following error when I try to spin this up locally:
> 
> ```
> app_1   | /usr/local/lib/ruby/site_ruby/2.6.0/rubygems.rb:275:in `find_spec_for_exe': Could not find 'bundler' (1.17.2) required by your /app/Gemfile.lock. (Gem::GemNotFoundException)
> app_1   | To update to the latest version installed on your system, run `bundle update --bundler`.
> app_1   | To install the missing version, run `gem install bundler:1.17.2`
> app_1   |       from /usr/local/lib/ruby/site_ruby/2.6.0/rubygems.rb:294:in `activate_bin_path'
> app_1   |       from /usr/local/bundle/bin/bundle:23:in `<main>'
> solr_1  | OpenJDK 64-Bit Server VM warning: Option UseConcMarkSweepGC was deprecated in version 9.0 and will likely be removed in a future release.
> blacklight_app_1 exited with code 1
> ```

Could this be because you have an outdated `Gemfile.lock` in your blacklight directory? 

I've got to say, I'm not surprised to see this. The combination of the dockerized app + engine_cart makes for a very awkward pairing. I'm not sure whether this is a scenario we can or should fix with code changes, or whether to address it via documentation? (In which case, see open question above.) FWIW, I would also be OK with dumping the `Dockerfile` entirely and using docker-compose only for Solr in CI. I'm not sure what having the app image is getting us, compared against the potential confusion of documenting and maintaining it.

Thanks for pulling this down and spinning it up, @dkinzer!

dkinzer commented:
@mjgiarlo sorry I didn't realize blacklight doesn't version control their Gemfile.lock... I'll try removing that and retrying.

dkinzer commented:
@mjgiarlo one thing though is that when I run `bundle exec rake ci` locally I get a bunch of errors and part of this effort I believe is to allow contributors to run the blacklight tests locally.

mjgiarlo commented:
@dkinzer :speech_balloon: 
> @mjgiarlo one thing though is that when I run `bundle exec rake ci` locally I get a bunch of errors and part of this effort I believe is to allow contributors to run the blacklight tests locally.

Are you doing this *while* the app container is running? Or afterward? And can you share the errors? Thx.

mejackreed commented on .env:
```diff
@@ -0,0 +1,5 @@
+RAILS_VERSION=5.2.4.1
+SOLR_PORT=8983
+SOLR_URL=http://solr:8983/solr/blacklight-core
+SOLR_VERSION=7
```
Is it possible to just run here with `latest` instead? That would be similar to the approach that SolrWrapper used.

mjgiarlo replied:
@mejackreed Yep! I'll make that change.

mejackreed commented:
Sorry adding one more thing I noticed in another comment. Strange, when I run `rails s` seperately I get the error 500 page as if it is running in `RAILS_ENV=production`. Not sure what is happening there but happy to keep digging. 👏 thanks again @mjgiarlo for moving this forward.

dkinzer commented:
@mjgiarlo I ran `bundle exec rake ci`  a second time after first making sure I first ran `docker-compose down` and also completely deleted the local .inernal_test_app folder and this time all tests passed with no errors. 🤷‍♂

The errors I ran into earlier were all related to solr not being available

```
 Autocomplete suggestions search form has suggest path
       Failure/Error: raise Blacklight::Exceptions::ECONNREFUSED, "Unable to connect to Solr instance using #{connection.inspect}: #{e.inspect}"

       Blacklight::Exceptions::ECONNREFUSED:
```

dkinzer commented:
FYI, we were having some webpacker issues on our docker instance and moving to node 8.x fixed it for us.

cdmo commented on .travis.yml:
```diff
@@ -35,16 +33,17 @@ notifications:
   email:
       - blacklight-commits@googlegroups.com
```
A bit unrelated to this PR, but are either of these notifications used? Maybe we should just delete them?

```suggestion
```

mjgiarlo replied:
will do!

cdmo commented:
Did I miss something, what's the `webpacker` issue? Or is this a reference to the feature specs not working thing?

cdmo commented:
> The combination of the dockerized app + engine_cart makes for a very awkward pairing. I'm not sure whether this is a scenario we can or should fix with code changes, or whether to address it via documentation? (In which case, see open question above.) FWIW, I would also be OK with dumping the Dockerfile entirely and using docker-compose only for Solr in CI. 

I was going to ask about this. In Travis engine_cart runs the various versions of ruby and rails. Locally it does too. Will this PR replace engine_cart for local and just test against one version of ruby? Thanks for all the contributions and conversations.

dkinzer commented:
I think this PR should be merged ASAP since the current state of the docker environment on the master branch is broken.

We can iterate on the merged changes in the future if need be.

dkinzer commented:
@cdmo yes, that was a reference to the features specs not working (when testing locally)... I'm pretty sure issue happens because app fails to load due to webpacker problems.   Also we are not seeing the errors that cause this because they are being hidden.

If I have time later on today, I'll try to push a PR against this branch.

mjgiarlo commented:
@mejackreed 
> 
> > It is not clear to me where information like this is currently documented. The README is short and sweet, and for this reason I suspect it is not a good place for such internal development documentation to live. Other options include CONTRIBUTING.md, but that is more focused on higher-level contribution process than particular technical steps.
> 
> Maybe we could/should just add a new wiki page for Blacklight development contribution?
> 

I renamed an existing page, updated its link on the main wiki page, and will use it to stash this information: https://github.com/projectblacklight/blacklight/wiki/Testing-and-Developing-Blacklight

:cool: :question:

mjgiarlo commented:
@cdmo :speech_balloon: 
> 
> I was going to ask about this. In Travis engine_cart runs the various versions of ruby and rails. Locally it does too. Will this PR replace engine_cart for local and just test against one version of ruby? Thanks for all the contributions and conversations.

No, I don't think it will. That's part of why it's an awkward pairing. But I believe the docker componentry does give you the ability to specify Ruby and Rails versions to do whatever you'd like. For a working set of steps, here's what I just did:

https://github.com/projectblacklight/blacklight/wiki/Testing-and-Developing-Blacklight

mjgiarlo commented:
@mejackreed can you try again following the guide I wrote up here: https://github.com/projectblacklight/blacklight/wiki/Testing-and-Developing-Blacklight

mjgiarlo commented:
May I ask @jcoyne and @cdmo for a final review?

cdmo commented:
Heh 😅 I dont know if I really ought to be reviewing this, kinda just wanted to see how I could help / be a tester. It all seems fine to me I suppose! I am checking out learndocker.online, thanks again for that.

I ran the wiki page instructions and hit a couple minor snags. First, running just `rake` and `rspec` lead to problems for me because I have multiple versions of these gems locally, specifying `bundle exec` as a prefix worked fine and got the right version. I also had one failing spec, sadly a spec I myself wrote (maybe the only one!). `facets_spec.rb:83` the expectation's left side was `nil`, as in the focus was not set as expected. I did confirm that in an actual FF browser the focus setting does work there. Spent some time trying to figure out why that's happening, but I think it's safe to say it's unrelated to what you are working on!

mjgiarlo commented:
@cdmo :speech_balloon: 
> Heh  I dont know if I really ought to be reviewing this, kinda just wanted to see how I could help / be a tester. It all seems fine to me I suppose! I am checking out learndocker.online, thanks again for that.

OK, no problem. Cool!
 
> I ran the wiki page instructions and hit a couple minor snags. First, running just `rake` and `rspec` lead to problems for me because I have multiple versions of these gems locally, specifying `bundle exec` as a prefix worked fine and got the right version. I also had one failing spec, sadly a spec I myself wrote (maybe the only one!). `facets_spec.rb:83` the expectation's left side was `nil`, as in the focus was not set as expected. I did confirm that in an actual FF browser the focus setting does work there. Spent some time trying to figure out why that's happening, but I think it's safe to say it's unrelated to what you are working on!

I just added some text to the Testing & Developing Blacklight wiki page about prefixing with `bundle exec` if needed. (I am still a big `rvm` user, so I never use `bundle exec`, and this will vary across developers.) Thanks for that feedback!

No idea about the failing spec. That's weird that it doesn't also fail in CI. Agree w/ you that it's unrelated to this PR, at least.

mjgiarlo commented:
Let's wait for one more :heavy_check_mark: before merging. @jcoyne :question:
